src/coreclr/src/utilcode/pedecoder.cpp

```diff
@@ -2319,22 +2319,95 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
-        if (pHeader->Signature == READYTORUN_SIGNATURE)
+        IMAGE_DATA_DIRECTORY* pDir = &GetCorHeader()->ManagedNativeHeader;
+
+        if (VAL32(pDir->Size) >= sizeof(READYTORUN_HEADER) && CheckDirectory(pDir))
         {
-            const_cast<PEDecoder *>(this)->m_pReadyToRunHeader = pHeader;
-            return pHeader;
+            PTR_READYTORUN_HEADER pHeader = PTR_READYTORUN_HEADER((TADDR)GetDirectoryData(pDir));
+            if (pHeader->Signature == READYTORUN_SIGNATURE)
+            {
+                const_cast<PEDecoder*>(this)->m_pReadyToRunHeader = pHeader;
+                return pHeader;
+            }
         }
     }
+    
+    READYTORUN_HEADER *nativeReadyToRunHeader = FindNativeReadyToRunHeader();
+    if (nativeReadyToRunHeader != NULL)
+    {
+        return nativeReadyToRunHeader;
+    }
 
     const_cast<PEDecoder *>(this)->m_flags |= FLAG_HAS_NO_READYTORUN_HEADER;
     return NULL;
 }
 
+READYTORUN_HEADER *PEDecoder::FindNativeReadyToRunHeader() const
+{
+#ifndef DACCESS_COMPILE
+    // Get the export directory entry
+    PIMAGE_DATA_DIRECTORY exportDirectoryEntry = GetDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT);
+    if (exportDirectoryEntry->VirtualAddress == 0 || exportDirectoryEntry->Size == 0)
+    {
+        return NULL;
+    }
+    
+    const uint8_t *imageBase = (const uint8_t *)GetBase();
+    uint32_t exportDirectory = exportDirectoryEntry->VirtualAddress;
+    
+    uint32_t namePointerCount;
+    uint32_t addressTableRVA;
+    uint32_t namePointersRVA;
+
+    if (!ReadUInt32(exportDirectoryEntry->VirtualAddress + 0x18, namePointerCount) || // +0x18: number of name pointers
+        !ReadUInt32(exportDirectoryEntry->VirtualAddress + 0x1C, addressTableRVA) || // +0x1C: export address table RVA
+        !ReadUInt32(exportDirectoryEntry->VirtualAddress + 0x20, namePointersRVA)) // +0x20: name pointer RVA
+    {
+        return NULL;
+    }
+
+    for (uint32_t nameIndex = 0; nameIndex < namePointerCount; nameIndex++)
+    {
+        uint32_t namePointerRVA;
+        if (ReadUInt32(namePointersRVA + sizeof(uint32_t) * nameIndex, namePointerRVA) &&
+            namePointerRVA != 0 &&
+            namePointerRVA + 11 <= GetSize()) // we need at least 11 bytes to encode the string RTR_HEADER and the terminating null
+        {
+            const char *namePointer = (const char *)&imageBase[namePointerRVA];
+            if (!strcmp(namePointer, "RTR_HEADER"))
+            {
+                uint32_t headerRVA;
+                if (ReadUInt32(addressTableRVA + sizeof(uint32_t) * nameIndex, headerRVA))
+                {
+                    return (READYTORUN_HEADER *)&imageBase[headerRVA];
+                }
+            }
+        }
+    }
+#endif
+
+    return NULL;
+}
+
+bool PEDecoder::ReadUInt32(int offset, uint32_t& outValue) const
+{
+    outValue = 0;
+
+    if (offset < 0 || (uint32_t)offset + sizeof(int) > GetSize())
```

Comment:
I do not think you need to do bounds checks like this here. They do not actually guarantee that you won't crash.

I would just cast the pointer to `IMAGE_EXPORT_DIRECTORY` and be done with it.

Reply:
Simplified based on your suggestion in 9th commit.